File: credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>
#include <stdlib.h>

int validate(long n);
long get_cc();

int main(void)
{
   switch (validate(get_cc()))
    {
        case 34:
            printf("AMEX\n");
            break;
         case 37:
            printf("AMEX\n");
            break;
         case 40:
            printf("VISA\n");
            break;
        case 41:
            printf("VISA\n");
            break;
        case 42:
            printf("VISA\n");
            break;
        case 43:
            printf("VISA\n");
            break;
        case 44:
            printf("VISA\n");
            break;
        case 45:
            printf("VISA\n");
            break;
        case 46:
            printf("VISA\n");
            break;
        case 47:
            printf("VISA\n");
            break;
        case 48:
            printf("VISA\n");
            break;
        case 49:
            printf("VISA\n");
            break;
        case 51:
            printf("MASTERCARD\n");
            break;
        case 52:
            printf("MASTERCARD\n");
            break;
        case 53:
            printf("MASTERCARD\n");
            break;
        case 54:
            printf("MASTERCARD\n");
            break;
        case 55:
            printf("MASTERCARD\n");
            break;
        default:
            printf("INVALID\n");
    }

}


long get_cc()
{
    long n = 0;
    do
    {
        n = get_long("Enter a credit card number: ");
    }
    while (n < 0 || n > 50000000000000000);

    if (floor(log10(labs(n))) + 1 == 13 || floor(log10(labs(n))) + 1 == 15 || floor(log10(labs(n))) + 1 == 16)
    {
        printf("%f", floor(log10(labs(n))) + 1);
        return nan;
    }



    return n;



}

int validate(long n)
{

    long number = n;
    int sum = 0;

    //printf("Number: %ld\n", number);


    sum = number % 10;
    number = number / 10;
    while(number)
    {
        int temp = (number % 10) * 2;
        if (temp > 9)
        {
            temp = temp - 9;
        }
        sum += temp;
        number /= 10;
        temp = number % 10;
        sum += temp;
        number /= 10;
    }

    while (n > 100)
    {
        n = n / 10;
    }


    //printf("Sum: %d\nLeftmost: %ld\n", sum, n);

    if (floor(log10(labs(n))) + 1 != 13 || floor(log10(labs(n))) + 1 != 15 || floor(log10(labs(n))) + 1 != 16 || sum % 10 )
    {
        printf("%ld\n", floor(log10(labs(n))) + 1);
        return -1;
    }
    return n;
}

File: sentimental-credit/credit.py
import cs50

def validate(n):
    number = n
    sum = 0
    sum = number % 10
    number = number // 10  # use integer division

    while number:
        temp = (number % 10) * 2
        if temp > 9:
            temp = temp - 9

        sum += temp
        number //= 10
        temp = number % 10
        sum += temp
        number //= 10

    while n > 100:
        n = n // 10

    if sum % 10:
        return False

    return n

def get_cc():
    while True:
        try:
            card_number = int(input("Please enter a valid credit card number: "))
            if card_number > 0:
                if len(str(card_number)) in [13, 15, 16]:  # convert to string before len()
                    return card_number
        except ValueError:
            print("Invalid input. Please enter an integer.")

def switch(n):
    if n == 34 or n == 37:
        return "AMEX\n"
    elif 40 <= n < 50:
        return "VISA\n"
    elif 51 <= n < 56:
        return "MASTERCARD\n"
    else:
        return "INVALID\n"

card_number = get_cc()
valid_number = validate(card_number)
switch(valid_number)










